site/client/src/main/scala/Api.scala

```diff
@@ -18,6 +19,29 @@ import common.ExtAjax._
 import org.scalajs.dom.ext.{ Ajax, AjaxException }
 
 object Client {
+  def readProgress(library: String, section: String, raw: String): List[ClientExercise] = {
+    val parsedBody = read[LibrarySectionArgs](raw)
+    parsedBody.exercises.map(e ⇒ {
+      ClientExercise(
+        library = library,
+        section = section,
+        method = e.methodName,
+        arguments = e.args,
+        state = if (e.succeeded) Solved else Unsolved
+      )
+    })
+  }
+
+  def fetchProgress(library: String, section: String): Future[Option[List[ClientExercise]]] = {
+    val url = Routes.Exercises.progress(library, section)
+    Ajax.get(url).map(r ⇒ {
```

Comment:
This is essentially map + filter which may be simplified via collect and partial function.